File: Samples/Java/Console/src/main/java/io/cratis/chronicle/samples/console/CustomerReducer.java
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.chronicle.samples.console;

import io.cratis.chronicle.observation.Reducer;

@Reducer
public class CustomerReducer {

    // Eight positional arguments, and the compiler checks every one of them. With plain Strings,
    // handing the event's full name to the email parameter would have compiled and shipped.
    public CustomerDetails customerRegistered(CustomerRegistered event) {
        return new CustomerDetails(
            event.customerId(),
            event.fullName(),
            event.email(),
            event.phoneNumber(),
            new StreetAddress(""),
            new City(""),
            new PostalCode(""),
            new Country("")
        );
    }

    public CustomerDetails customerAddressUpdated(CustomerAddressUpdated event, CustomerDetails state) {
        CustomerDetails current = state != null ? state : new CustomerDetails();
        return new CustomerDetails(
            current.getId(),
            current.getFullName(),
            current.getEmail(),
            current.getPhoneNumber(),
            event.streetAddress(),
            event.city(),
            event.postalCode(),
            event.country()
        );
    }
}

File: Samples/Kotlin/Console/src/main/kotlin/io/cratis/chronicle/samples/console/CustomerReducer.kt
// Copyright (c) Cratis. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package io.cratis.chronicle.samples.console

import io.cratis.chronicle.observation.Reducer

@Reducer
class CustomerReducer {

    // Each value goes from the event to the read model without a conversion, and without anyone
    // checking that the name did not end up in the email. The types are the check.
    fun customerRegistered(event: CustomerRegistered): CustomerDetails {
        return CustomerDetails(
            id = event.customerId,
            fullName = event.fullName,
            email = event.email,
            phoneNumber = event.phoneNumber
        )
    }

    fun customerAddressUpdated(event: CustomerAddressUpdated, state: CustomerDetails?): CustomerDetails {
        return (state ?: CustomerDetails()).copy(
            streetAddress = event.streetAddress,
            city = event.city,
            postalCode = event.postalCode,
            country = event.country
        )
    }
}
